component/status.go

```diff
@@ -0,0 +1,106 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package component
+
+import (
+	"errors"
+	"time"
+)
+
+type Status int
+
+const (
+	StatusOK Status = iota
+	StatusError
+)
+
+// StatusSource component that reports a status about itself.
+// The implementation of this interface must be comparable to be useful as a map key.
+type StatusSource interface {
+	ID() ID
+}
```

Comment:
Why do you need this interface at all? Simply pass the ID?

Reply:
We need a key that uniquely identifies the component. ID is not a unique identification. You may have a processor and a receiver that have the same ID (we don't enforce that receivers and processors don't have the same type string - although in reality they are all unique in core and contrib).

Why it needs to be unique? So that the StatusWatcher can maintain a `map[StatusSource]Status`.

Alternatively we should enforce uniqueness of ID across all component types.

Reply:
Interesting point... I understand now the need, let me think a bit more, probably better to have this as a struct if we can define it.

Reply:
Yet another alternate is to require the Host to track the aggregate component status. In that case StatusWatcher will only be notified of that single aggregate bool status. But in that case we have to hard-code the aggregation logic in the Host and lose flexibility in allowing the extension to decide what to do with individual component statuses.

Reply:
For the "global component ID", I think we need something like:

```
package component

type GlobalID struct {
  ID ID
  Kind Kind
  PipelineID ID  // Not empty only if the Kind is Processor
}
```

Reply:
This also requires https://github.com/open-telemetry/opentelemetry-collector/pull/6540 otherwise you can still have duplicates even with GlobalID.

Reply:
Good point about that. Now I am more confused after I looked at your implementation https://github.com/open-telemetry/opentelemetry-collector/pull/6560/files#diff-bcb3fc758a54acf0dda21a0d9c79b6cf6e6adcf58d0a282c93be50553523b09dR41 and https://github.com/open-telemetry/opentelemetry-collector/pull/6560/files#diff-19abf8e6f60ec275051d1447d4d97f4c55b449192f2cb946e3e4ea44ff49fcc0R69

At least in the `statusReportingComponent` is not unique, so my suggestion at least improves that :))

Reply:
I think `&statusReportingComponent{}` is unique. We create a new instance for every created component. It is not a value of `statusReportingComponent{}` struct, we create a pointer to a new struct every time.

Reply:
Added a test that verifies the uniqueness of `StatusSource`: https://github.com/tigrannajaryan/opentelemetry-collector/blob/2ac7d388cf548e538601796ec9fff25ae5eb6a78/service/service_test.go#L229